### ISSUE TITLE
Add vDPA migration support

### ISSUE THREAD

**Issue by sboeuf** [cloud-hypervisor/cloud-hypervisor#3809]
vDPA is a kernel framework introduced fairly recently in order to handle devices complying with virtio specification on their datapath, while the control path is vendor specific. For the datapath, that means the virtqueues are handled through DMA directly between the hardware and the guest, while the control path goes through the vDPA framework, eventually exposed through a `vhost-vdpa` device.

vDPA, like VFIO, aims at achieving baremetal performance for devices that are passed into a VM. But unlike VFIO, it provides a simpler/better framework for achieving migration. Because the DMA accesses between the device and the guest are going through virtio queues, migration can be achieved way more easily, and doesn't require each device driver to implement the migration support. In the VFIO case, each vendor is expected to provide an implementation of the VFIO migration framework, which makes things harder as it must be done for each and every device.

So to summarize the point is to support migration for hardware devices through which we can achieve baremetal performances.

The way to implement this for Cloud Hypervisor should be fairly straightforward:
- [ ] Create a new type of device called `Vdpa` that will implement the `VirtioDevice` trait (skeleton) and will be placed on the `virtio-pci` bus
- [ ] Expose this device to the user with a new CLI/HTTP API option called `vdpa`
- [ ] Implement `Vdpa` by relying on the `rust-vmm/vhost` implementation of the `vhost-vdpa` interface
- [ ] Add integration tests relying on both `vdpa-sim-net` and `vdpa-sim-blk` kernel modules to simulate `net` and `blk` vDPA devices, which will be exposed as `virtio-net` and `virtio-block` devices in the guest.

**Comment by likebreath** [comment]
> For the datapath, that means the virtqueues are handled through DMA directly between the hardware and the guest, while the control path goes through the vDPA framework.

Sounds like our original concern of vDPA being less secure is actually not a problem, right?

**Comment by sboeuf** [comment]
> > For the datapath, that means the virtqueues are handled through DMA directly between the hardware and the guest, while the control path goes through the vDPA framework.
> 
> Sounds like our original concern of vDPA being less secure is actually not a problem, right?

Yes because the virtqueues are directly handled by the device, and the device is placed behind the host IOMMU to ensure it's not trying to access unallowed memory.

**Comment by sboeuf** [comment]
The support for vDPA has been merged through https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3849.

Here is the list of things left to do:
- [ ] Write documentation
- [ ] Migration support
- [ ] vIOMMU support

**Comment by sboeuf** [comment]
Giving some update about recent investigations and discovery about vDPA migration:

- `vhost-vdpa` doesn't support `VHOST_SET_LOG_BASE`, as there's currently no HW implementation allowing the dirty pages to be logged. There are some ideas about using the PF of the device to report these, or there might be some solution relying on PASID, but this requires recent IOMMU hardware, and the code isn't there yet in the Linux kernel, therefore we can't assume this is going to land anytime soon.
- `vhost-vdpa` doesn't support "pausing" the device. Usually in `vhost-user` and `vhost`, `VHOST_GET_VRING_BASE` is used to notify the backend it should stop processing the queues. If this was available, we could rely on it to implement snapshot/restore and a version of live migration that would stop the device before we send the guest memory in multiple rounds. The point is, even without a way to log dirty pages, we could simplify the design and still get migration to work with a small impact on the efficiency of the migration.
- An alternative for performing migration is already part of DPDK and is being submitted to QEMU. This is called SW assisted migration as it relies on the userspace driver (DPDK or VMM) to track which pages are dirtied. It uses a concept of shadowing the virtqueues. Drawback of this approach, it increases greatly the complexity of vDPA support, and reduces the baremetal performance we can expect from the vDPA device.
- From a Cloud Hypervisor's perspective, the best tradeoff if we wanted to support SW assisted migration would be to add a vDPA alternative to our `vhost-user-net` device. This way we could have the VMM rely on DPDK ability to migrate a vDPA device. This will only be evaluated if there's a strong need from one of our users.

To summarize, we're going to let vDPA grow a bit more and let the Linux community decide what is the best approach for HW migration support. Only after things will have settled, we'll re-assess what needs to happen in Cloud Hypervisor to support vDPA migration.

**Comment by sboeuf** [comment]
Starting with Linux v5.19, a new vhost ioctl `VHOST_VDPA_SUSPEND` has been added. It allows the consumer (the VMM) to inform the vDPA device to stop processing the virtqueues, which is useful in the context of both offline and online migration.
The `vdpa_sim` framework supports this vhost feature, meaning the CI can be updated to validate the migration support once it will be added.

**Comment by eugpermar** [comment]
Hi Sebastien,

While it's true that the only way to resume a vdpa device is to reset it for the moment it doesn't mean the guest is aware of this reset.

The vhost-vdpa device config space exposed to the guest (and, in particular, the status) is not directly the vdpa device's one, but a new emulated one. If the guest reads the state during this switch, the vmexit will wait for the BQL unlock, and then will read it's still with DRIVER_OK. All the resume is handled by qemu.

The guest can see that it has lose some in-flights packets however, but it is more or less expected from network. We're working on migrate in-flighs buffers too.

Thanks!

**Comment by eugpermar** [comment]
> > While it's true that the only way to resume a vdpa device is to reset it for the moment it doesn't mean the guest is aware of this reset.
> > The vhost-vdpa device config space exposed to the guest (and, in particular, the status) is not directly the vdpa device's one, but a new emulated one. If the guest reads the state during this switch, the vmexit will wait for the BQL unlock, and then will read it's still with DRIVER_OK. All the resume is handled by qemu.
> 
> Ah yes you're right the guest isn't made aware of the reset. But do you recommend using the `suspend`/`reset` operations for a simple VM pause/resume (no migration involved)?
> 

For net I'd say so. To drop some packets or connections is expected anyway, so we shouldn't lose a lot with this method.

Since the controversial point of block is how to get the in flight buffers, I'd say that a resume operation / ioctl would fit better. That could enable other devices that are currently non migratable (virtio-fs, crypto, etc) to also suspend and resume.

Thanks!

**Comment by sboeuf** [comment]
>Since the controversial point of block is how to get the in flight buffers, I'd say that a resume operation / ioctl would fit better. That could enable other devices that are currently non migratable (virtio-fs, crypto, etc) to also suspend and resume.

Is there any effort to push such thing to the vhost uAPI? TBH I'm surprised only `suspend` operation has been added. Usually these operations are defined as pairs (like start/stop, play/pause, etc...).

**Comment by sboeuf** [comment]
> > > Since the controversial point of block is how to get the in flight buffers, I'd say that a resume operation / ioctl would fit better. That could enable other devices that are currently non migratable (virtio-fs, crypto, etc) to also suspend and resume.
> > 
> > 
> > Is there any effort to push such thing to the vhost uAPI? TBH I'm surprised only `suspend` operation has been added. Usually these operations are defined as pairs (like start/stop, play/pause, etc...).
> 
> No that I know of, it just simplified things reducing the number of possible transitions between states.

Ok!

**Comment by sboeuf** [comment]
@eugpermar I've just realized that using `suspend`/`reset` is really not appropriate for the `pause`/`resume` use case. The problem comes from the `reset`, which implicitly invalidates all IOTLB entries. That means for a simple `resume`, we'd end up reinitializing the entire device. Of course this is not a problem for migration, but for snapshot/restore or pause/resume this is not acceptable IMO.
Having a resume ioctl would make lots of sense.

**Comment by eugpermar** [comment]
> @eugpermar I've just realized that using `suspend`/`reset` is really not appropriate for the `pause`/`resume` use case. The problem comes from the `reset`, which implicitly invalidates all IOTLB entries. That means for a simple `resume`, we'd end up reinitializing the entire device. Of course this is not a problem for migration, but for snapshot/restore or pause/resume this is not acceptable IMO. Having a resume ioctl would make lots of sense. Do you think I could push for such change?

Sure. It was actually proposed by previous versions of the patch, like [[v5,1/4] vdpa: Add suspend operation](https://patchwork.kernel.org/project/kvm/patch/20220603100944.871727-2-eperezma@redhat.com/). To remove it just simplified things at that moment.

Since this new resume capability would require another feature bit, I don't find any impediment for it as long as its semantic is well defined. And it would enable new use cases for some non migratable devices.

Maybe the semantics for suspend need to be expanded if the device have the resume ioctl, for example retaining in flight buffers.

**Comment by sboeuf** [comment]
> > @eugpermar I've just realized that using `suspend`/`reset` is really not appropriate for the `pause`/`resume` use case. The problem comes from the `reset`, which implicitly invalidates all IOTLB entries. That means for a simple `resume`, we'd end up reinitializing the entire device. Of course this is not a problem for migration, but for snapshot/restore or pause/resume this is not acceptable IMO. Having a resume ioctl would make lots of sense. Do you think I could push for such change?
> 
> Sure. It was actually proposed by previous versions of the patch, like [[v5,1/4] vdpa: Add suspend operation](https://patchwork.kernel.org/project/kvm/patch/20220603100944.871727-2-eperezma@redhat.com/). To remove it just simplified things at that moment.
> 
> Since this new resume capability would require another feature bit, I don't find any impediment for it as long as its semantic is well defined. And it would enable new use cases for some non migratable devices.

It's actually making the logic more complex now that we'd have one feature bit for suspend and one for resume. But I guess it's up to us to clearly define that devices which support suspend only are meant for live migration while devices supporting both suspend and resume can be used for pausing/resuming a VM, and therefore provide some offline migration (snapshot/restore).

> 
> Maybe the semantics for suspend need to be expanded if the device have the resume ioctl, for example retaining in flight buffers.

Yes I think if the device has both ability to suspend and resume, then it shouldn't drop anything when suspended. The point is simply to prevent it from processing virtqueues.

**Comment by eugpermar** [comment]
> > > @eugpermar I've just realized that using `suspend`/`reset` is really not appropriate for the `pause`/`resume` use case. The problem comes from the `reset`, which implicitly invalidates all IOTLB entries. That means for a simple `resume`, we'd end up reinitializing the entire device. Of course this is not a problem for migration, but for snapshot/restore or pause/resume this is not acceptable IMO. Having a resume ioctl would make lots of sense. Do you think I could push for such change?
> > 
> > 
> > Sure. It was actually proposed by previous versions of the patch, like [[v5,1/4] vdpa: Add suspend operation](https://patchwork.kernel.org/project/kvm/patch/20220603100944.871727-2-eperezma@redhat.com/). To remove it just simplified things at that moment.
> > Since this new resume capability would require another feature bit, I don't find any impediment for it as long as its semantic is well defined. And it would enable new use cases for some non migratable devices.
> 
> It's actually making the logic more complex now that we'd have one feature bit for suspend and one for resume. But I guess it's up to us to clearly define that devices which support suspend only are meant for live migration while devices supporting both suspend and resume can be used for pausing/resuming a VM, and therefore provide some offline migration (snapshot/restore).
> 

A little bit more complex for the guest part, having to check two feature flags, but enabling the possibility for vendors that does not want or can implement the resume part.

> > Maybe the semantics for suspend need to be expanded if the device have the resume ioctl, for example retaining in flight buffers.
> 
> Yes I think if the device has both ability to suspend and resume, then it shouldn't drop anything when suspended. The point is simply to prevent it from processing virtqueues.

Right.
